pkg/cloudprovider/vm.go

```diff
@@ -130,33 +126,69 @@ type vmIdentityInfo struct {
 	info *compute.VirtualMachineIdentity
 }
 
-func (i *vmIdentityInfo) RemoveUserIdentity(id string) error {
-	if err := filterUserIdentity(&i.info.Type, i.info.IdentityIds, id); err != nil {
-		return err
-	}
-	// If we have either no identity assigned or have the system assigned identity only, then we need to set the
-	// IdentityIds list as nil.
-	if i.info.Type == compute.ResourceIdentityTypeNone || i.info.Type == compute.ResourceIdentityTypeSystemAssigned {
-		i.info.IdentityIds = nil
+func (i *vmIdentityInfo) GetUserIdentityList() []string {
+	var ids []string
+	if i.info == nil {
+		return ids
 	}
-	// if the identityids is nil and identity type is not set, then set it to ResourceIdentityTypeNone
-	if i.info.IdentityIds == nil && i.info.Type == "" {
-		i.info.Type = compute.ResourceIdentityTypeNone
+	for id := range i.info.UserAssignedIdentities {
+		ids = append(ids, id)
 	}
-	return nil
+	return ids
 }
 
-func (i *vmIdentityInfo) AppendUserIdentity(id string) bool {
-	if i.info.IdentityIds == nil {
-		var ids []string
-		i.info.IdentityIds = &ids
+func (i *vmIdentityInfo) SetUserIdentities(ids map[string]bool) bool {
+	expectedIDList := make(map[string]bool)
+
+	if i.info.UserAssignedIdentities == nil {
+		i.info.UserAssignedIdentities = make(map[string]*compute.VirtualMachineIdentityUserAssignedIdentitiesValue)
 	}
-	return appendUserIdentity(&i.info.Type, i.info.IdentityIds, id)
-}
 
-func (i *vmIdentityInfo) GetUserIdentityList() []string {
-	if i.info.IdentityIds == nil {
-		return []string{}
+	for id, _ := range i.info.UserAssignedIdentities {
+		expectedIDList[id] = true
+	}
+
+	for id, add := range ids {
+		_, exists := expectedIDList[id]
+
+		// identity already exists/not exists, so skip doing add/del operation for
+		// the identity again for the node
+		if (exists && add) || (!exists && !add) {
+			delete(ids, id)
+		}
+		// already exists on node and want to remove existing identity
+		if exists && !add {
+			delete(expectedIDList, id)
+		}
+		// doesn't exist on the node and want to add new identity
+		if !exists && add {
+			expectedIDList[id] = true
+		}
+	}
+
+	// all identities are the node are to be removed
+	if len(expectedIDList) == 0 {
+		i.info.UserAssignedIdentities = nil
+		if i.info.Type == compute.ResourceIdentityTypeSystemAssignedUserAssigned {
+			i.info.Type = compute.ResourceIdentityTypeSystemAssigned
+		} else {
+			i.info.Type = compute.ResourceIdentityTypeNone
+		}
+		return true
 	}
-	return *i.info.IdentityIds
+	// add and remove the new list of identities keeping the same type as before
+	userAssignedIdentities := make(map[string]*compute.VirtualMachineIdentityUserAssignedIdentitiesValue)
```

Comment:
Looks like we can have common code for this and the code in vmss with just the change of the type being used to assign to the value ? Can we make that change, please ?